### ISSUE TITLE
Handling of regex patterns that can match the empty string

### ISSUE THREAD

**Issue by simark** [BurntSushi/ripgrep#729]
A regex that can match an empty string, like this:

    rg '(thisstringdoesnotexist)?'

makes rg output one match for every line of every file searched.  With the --vimgrep option, it's one match per character.  When searching a big directory, it makes rg take a lot of time and resources to output meaningless results.  I think that rg should not output matches that are zero-length, I can't think of a use case where it would be useful.

**Comment by okdana** [comment]
There is one use case i know of: If you want to colourise matches, but don't want to restrict the output to the lines containing those matches, it's common (with `grep` too) to do something like `rg '^|foo'`. That way every instance of `foo` is coloured, but it prints all of the 'non-matching' lines too

I suppose `rg -C65535 foo` would be a viable alternative to that though

**Comment by simark** [comment]
Ah ok thanks, didn't know about that use case...

**Comment by simark** [comment]
> Edit: In your case it just seems like a weird pattern though. Why would you search for (thisstringdoesnotexist)? instead of just thisstringdoesnotexist?

Indeed, it would not be a useful search pattern.  However, I am integrating rg in an IDE, and am trying to handle gracefully anything the user throws at us, good or bad.
